### ISSUE TITLE
Proposal: our expectation on KubeFlow

### ISSUE THREAD

**Issue by DjangoPeng** [google/kubeflow#33]
**Authors:**

- **@DjangoPeng** - Jingtian Peng \<pengjingtian@caicloud.io\>
- **@zjx-caicloud** - Jianxin Zhang \<zhangjianxin@caicloud.io\>
- **@gaocegege** - Ce Gao \<gaoce@caicloud.io\>
- **@ScorpioCPH** - Penghao Cen \<cenph@caicloud.io\>

### Motivation

TensorFlow users can run training jobs on Kubernetes with the support of this project: [tensorflow/k8s](https://github.com/tensorflow/k8s), which is a CRD controller (or TensorFlow operator) for deploying distributed TensorFlow on Kubernetes.

Maybe it is not sustainable for TensorFlow users (e.g. data scientist or deep learning engineer
) to write a complex YAML file to deploy their TensorFlow Training and other related service (e.g. launch TensorFlow Serving and TensorBoard). And the users could not get the training status in the current design.

The KubeFlow project is dedicated to making a high-efficient, easy-to-use, cloud-native and scalable Machine Learning platform based on Kubernetes. Our goal is building an **end-to-end** system to **manage the whole lifecycle of Machine Learning Jobs**, which makes the TensorFlow to be kubernetes-friendly and user-friendly.

This doc describes a user-oriented distributed ML platform with:

- Defining a user spec to deploy TensorFlow Training/Serving/TensorBoard job
- Workflow from user side
- Components of the architecture

### Use Cases

- I have no idea about Kubernetes and just want to focus on developing TensorFlow  model.
- I don’t need to write YAML files for any kind of TensorFlow jobs, including training, serving and TensorBoard jobs.
- I can specify computing resources (e.g. GPUs) which will be used in TensorFlow training or serving.
- I can get the processing status and logs from KubeFlow API after triggering TensorFlow training jobs.
- I can launch TensorFlow model as a service to expose trained TensorFlow model.
- I can launch TensorBoard as a service which would track multiple experiments.

### Goals

- Implement a server
  - Accept training code files, e.g. python files as input, and run training jobs on Kubernetes, with no need for additional configs.
  - watch job status running on Kubernetes and do some cleanup works on the cluster
- Implement a monitor, to get the job status from Kubernetes and report specific status to the KubeFlow server
- Update CRD specification and controller to fit the new design
- For scheduling (future work):
  - Implement customized scheduler or reuse the community project [kubernetes-incubator/kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator) to support fine-grained scheduling for TensorFlow jobs
- For resource management (future work):
  - Use device plugin (in Kubelet) and NVIDIA [k8s-device-plugin](https://github.com/NVIDIA/k8s-device-plugin) with nvidia-docker v2 for NVIDIA GPU scheduling
  - Implement Resource Class with CRD and Custom Scheduler for different type GPUs scheduling (on homogeneous nodes)

### Workflow

![kubeflow_training](https://user-images.githubusercontent.com/16943353/33839179-88469962-decc-11e7-8483-42e8054e7b09.png)

**Fig 1** KubeFlow distributed training job

As shown in Fig 1, users develop some ML models in the JupyterHub. Then they send the TensorFlow Python script (e.g. inception_v3.py) into KubeFlow.  When receiving the script, KubeFlow creates TensorFlow distributed training job based on Kubernetes. Finally, the TensorFlow training job would delivery sort of files, such as tf.events for TensorBoard visualization and saved_model.pb for TensorFlow Serving.

![kubeflow_serving](https://user-images.githubusercontent.com/16943353/33839189-8d36af48-decc-11e7-80d6-54a9053164ad.png)

**Fig 2** KubeFlow serving job

As shown in Fig 2, users send the saved_model.pb file into KubeFlow. Then KubeFlow creates a TensorFlow Serving job to provide a public service. With the number of requests increasing, KubeFlow would automatically scale out the job.

![kubeflow_tensorboard](https://user-images.githubusercontent.com/16943353/33839202-93c33c14-decc-11e7-9e9f-8dc7c09dea66.png)

**Fig 3** KubeFlow TensorBoard job

As shown in Fig 3, users send the tf.events file into KubeFlow. Then KubeFlow creates a TensorBoard job to track multiple experiments. 

### Components

![kubeflow_components](https://user-images.githubusercontent.com/16943353/33839209-97e32fac-decc-11e7-9767-471d56859b94.png)

**Fig 4** KubeFlow components

The design has several components:

- Kubernetes Extended (to better serve Machine Learning workload)
  - TensorFlow CRD
  - TensorFlow Controller
  - TensorFlow Scheduler
- Server (to manage TensorFlow jobs and interact with the client)
- Monitor (to watch and report statuses of all TensorFlow jobs)

#### CRD

According to the use case and design goal, TensorFlow jobs are always divided into three types as below:

- Training Job
  - Local Training Job
  - Distributed Training Job
- Serving Job
- TensorBoard Job

TensorFlow distributed training job is most complicated and elusive among them. It adopts the classical PS-Worker distributed framework. PS (Parameter Server) is responsible for storing and updating the model's parameters. Worker is responsible for computing and applying gradients.

But, the native implementation of TensorFlow distributed framework is not perfect. For example, PS processes would join forever even though all Workers processes completed their computation works, which is a waste of resource. At the same time, Kubernetes has no idea of the TenforFlow job for lack of TensorFlow conceptions and semantics. So, it's difficult for native Kubernetes to make a realtime monitor watching the TenforFlow job status. In this case, KubeFlow aims to handle all problems above and fill in the gap between TensorFlow and Kubernetes.

The state-of-art implementation tensorflow/k8s generally works well and lots of works could be reused in the new CRD design. We have some changes based on [tensorflow/k8s](https://github.com/tensorflow/k8s). We divides TensorFlow jobs into three types: training jobs, serving jobs and TensorBoard jobs.

In general, we always launch more than one TensorFlow training jobs with multiple kinds of hyper parameters combinations to get the best training results. At the same time, we can use one TensorBoard instance to visualize all of TensorFlow events outputted from training jobs. It is really a common case for data scientists and algorithm engineers. But, in the current design, the TensorBoard job is binding to one TensorFlow job.

In order to make above come true, we need to decouple the TensorFlow job and TensorBoard job. We can make some changes on the TenforFlow job CRD. At first, both jobs can share TFReplicaSpec. Moreover, we can use TFReplicaSpec to specify the TensorFlow Serving job. At the same time, we can add a type field to define the TensorFlow job type (one of Training, TensorBoard or Serving).

```
apiVersion: "tensorflow.org/v1alpha1"
kind: "TFJob"
metadata:
  name: "training-job"
spec:
  type: Training
  tfReplicaSpec:
  - replicas: 2
    tfReplicaType: PS
    template:
      spec:
        containers:
        - name: tensorflow
          image: tensorflow/tensorflow:1.4.0
          command:
          - "/workdir/tensorflow/launch_training.sh"
          volumeMounts:
          - name: workdir
            mountPath: /workdir
        volumes:
        - name: workdir
          glusterfs:
            endpoints: <gluster-cluster>
            path: <gluster_vol_subpath>
        restartPolicy: OnFailure
      tfReplicaSpec:
  - replicas: 4
    tfReplicaType: Worker
    template:
      spec:
        containers:
        - name: tensorflow
          image: tensorflow/tensorflow:1.4.0
          command:
          - "/workdir/tensorflow/launch_training.sh"
          args:
          - "--data_dir=/workdir/data"
          - "--train_dir=/workdir/train"
          - "--model_dir=/workdir/model"
          volumeMounts:
          - name: workdir
            mountPath: /workdir
        volumes:
        - name: workdir
          glusterfs:
            endpoints: <gluster-cluster>
            path: <gluster_vol_subpath>
        restartPolicy: OnFailure
```
**Algo 1** YAML config for distributed training job

Related issues: [tensorflow/k8s#209](https://github.com/tensorflow/k8s/issues/209)

#### Controller

[tensorflow/k8s](https://github.com/tensorflow/k8s) follows the CoreOS operator pattern to handle the events from Kubernetes, we aim to refactor the project to controller pattern which is similar to [kubernetes/sample-controller](https://github.com/kubernetes/sample-controller) to be extensible and robust.

But it has the same function as [tensorflow/k8s](https://github.com/tensorflow/k8s): the controller watches the shared state of the cluster through the Kubernetes apiserver and makes changes attempting to move the current state towards the desired state for TensorFlow jobs on Kubernetes.

#### Server

Server is the most important component in KubeFlow, which are responsible for:

Setting up RESTful apiserver to listen for requests for creating TensorFlow jobs or requests from KubeFlow monitor for status updating.

Generating objects that can be accepted by Kubernetes apiserver to create TensorFlow jobs according to user requests.

Recording status of TensorFlow jobs in order to reschedule TensorFlow jobs when something goes wrong during training.

#### Monitor

KubeFlow controller allows users to create TensorFlow training jobs on Kubernetes, but the parameter servers are always running even after the training job finished. Besides this, users could not get the Tensorflow job's' status unless they access to Kubernetes.

Then we aim to implement a component which is called KubeFlow monitor, to watch the status of TensorFlow jobs on Kubernetes. And the monitor should report the status to KubeFlow server, then the server could move the current state towards the desired state.

## Google Doc Version

For continuously and conveniently discuss the proposal and scope of KubeFlow, we created a [gdoc](https://docs.google.com/document/d/1dmErPUmqqKMOe4L0ZHQglSdgDguCM4SzlsEdYXRMIDA/edit?usp=sharing) of this proposal. Welcome for suggestion and comments! 🎉

**Comment by DjangoPeng** [comment]
This is our expectation on KubeFlow, including some use cases we are facing with and some thoughts about the scope of KubeFlow. We hope we can work with all the contributors in the KubeFlow community to make a fancy and ease-to-use Machine Learning system on Kubernetes. 

Welcome for comments and discussion. 🎉

**Comment by jlewi** [comment]
Thanks for starting the discussion

Some high level points

1. I think Kubeflow wants to support frameworks other than TensorFlow. 

1.  What do you mean by Kubernetes Agnostic? I think Kubeflow aims to be Kubernetes native. 
   * We want to depend on APIs/Conventions defined by Kubernetes to create a portable stack that can run anywhere Kubeflow runs
   * We want to be able to assume we are running on K8s so that we can depend on K8s elements
        * e.g. we want to be able to depend on an APIServer

In the spirit of a micro-service architecture, can we split this proposal into proposals for separate services/components? I agree Kubeflow definitely needs a way to track the history of training jobs and facilitate comparing the resulting models. (Running TensorBoard as part of the TfJob was not intended to solve that problem). 

Can we treat creating components/services for tracking models and comparing them as separate from serving and managing the low level details of training jobs? I think tracking trained models requires more than just launching TensorBoard. At a minimum, I expect we probably need a 
    * DB which keeps track of all the jobs
    * A UI which displays the jobs
    * A way to launch various analysis tools (e.g. TensorBoard) for a particular model 
    * An API that allows loading/querying data into the DB.

Ideally we'd leverage K8s patterns to do this in a convenient way e.g.
    * Associate labels with entities to facilitate querying
    * Use K8s patterns like admission controllers to monitor various types of resources and automatically create entries in the DB when certain objects appear (e.g. TfJob).

**Comment by erikerlandson** [comment]
@jlewi hopefully I am not reading too much into "DB" but I'm a bit wary of on-boarding a database unless there is a compelling use case for that kind of heavy-weight state.

**Comment by ScorpioCPH** [comment]
> Kubernetes-agnostic

I think this is from user side, it means that kubeflow users don't need to know the details about kubernetes, just focus on the TF model. In another word, it is user-friendly.

**Comment by DjangoPeng** [comment]
@jlewi Hi, thanks for your comments. 

> 1. I think Kubeflow wants to support frameworks other than TensorFlow.

Totally agree with you. In the Components fig, we make a Kubernetes Extended module and the supporting of all other ML frameworks could been implemented here. For example, we can implement the MXNet CRD/Controller/Scheduler to support MXNet framework. At the same time, they can share the common KubeFlow Server and Monitor.

> 2. What do you mean by Kubernetes Agnostic? 

As @ScorpioCPH said, it's from user side. And I've replace `Kubernetes-agnostic` with `user-friendly` for disambiguation.

> - We want to depend on APIs/Conventions defined by Kubernetes to create a portable stack that can run anywhere Kubeflow runs
> - We want to be able to assume we are running on K8s so that we can depend on K8s elements
>   - e.g. we want to be able to depend on an APIServer

Sure thing, and that's how we implement it. The purpose of this doc is explaining for both data scientist and other ML engineer without Kubernetes knowledges. For Kubernetes and KubeFlow developers, we would open individual proposal to exactly explain each component of KubeFlow.
It should be clear that KubeFlow Server is not alternative to Kubernetes apiserver. On the contrary, it's an user of Kubernetes apiserver.  KubeFlow Server is dedicated to handle all ML conceptions and semantics, and convert the ML request (e.g.Train a TensorFlow model) as a Kubernetes-friendly request.

**Comment by pineking** [comment]
@DjangoPeng thanks for your proposal. that's what we need and want to implement.
 
As a ML engineer, some questions:
1. do we plan to provide a command line tool to manage the training task (create/delete/monitor)?
2. I think that's a good idea user can create training task in one command, instead of writing YAML file, we also have implemented this.
3. it's necessary that KubeFlow server can support GPU Quota and Queue mode in multi-users case.
4. why do we not use node label and selector for different type GPU scheduling?

**Comment by jlewi** [comment]
@DjangoPeng @zjx-caicloud @gaocegege @ScorpioCPH 

Thank you so much for taking the time to put this together.  There's a lot of detail and thought in this proposal. What can I do to help you?

**Comment by gaocegege** [comment]
@jlewi @erikerlandson 

The KubeFlow server needs a DB to keep track of the lifecycle of TensorFlow jobs, but I am not sure if we should offer APIs for it.

We will file more detailed proposals for the design soon, and we could discuss this question there. :thinking:

**Comment by DjangoPeng** [comment]
@pineking Nice to meet you here. 

> 1. do we plan to provide a command line tool to manage the training task (create/delete/monitor)?

We have a internal plan on the client (e.g. CLI, Web UI), and it's open to discuss whether client is the scope of KubeFlow.

> 2. I think that's a good idea user can create training task in one command, instead of writing YAML file, we also have implemented this.

The requirement is indeed exist both from inside and outside. Hope we can work together on the fancy project.

When it comes to GPU supporting and scheduling, I think it might the scope of Kubernetes itself.

WDYT @ScorpioCPH

**Comment by ScorpioCPH** [comment]
@pineking 

Hi, thanks for your input:

> 3. it's necessary that KubeFlow server can support GPU Quota and Queue mode in multi-users case.

Sorry, i'm not very clear about `Queue mode`, do you mean if multi-users submit TF training jobs at the same time, should we put these jobs into a `Work Queue`, then launch them one by one?

And about `GPU Quota` resource management/scheduling, I think it is Kubernetes' responsibility to handle this with `device-plugin` feature (which is released in v1.8)

> 4. why do we not use node label and selector for different type GPU scheduling?

This is another topic `Resource-Class`, here is [proposal](https://github.com/kubernetes/community/pull/782).
I think you are on the right point, and i'm working on this idea, you can take a look at this if you're interested:
- design proposal [draft](https://github.com/caicloud/oscenter-docs/pull/1)
- resource controller [PR](https://github.com/caicloud/resource-controller/pull/1)
- resource scheduler [PR](https://github.com/caicloud/resource-scheduler/pull/1)

**Comment by jlewi** [comment]
I think a Web UI and CLI are within scope for Kubeflow.

@wbuchwalter already has an initial PR tensorflow/k8s#125 for a UI which I expect to be merged as soon as he gets back from traveling and he has time to resolve the conflicts.

Some high level points

* At the lowest level of Kubeflow I think we should avoid hiding K8s APIs or trying to provide simpler abstractions than K8s APIs
    * As an example, TfJob includes PodTemplate spec's rather than trying to define a new API to specify the process to run.
    * One advantage of using K8s APIs is that they are already familiar to K8s users
    * A second advantage, is that you get the full richness of K8s APIs
         * e.g. With TfJob you can leverage K8s volumes to manage storage, or specify any resource request/limit that K8s understands.
* A consequence of the above, is that the lowest level APIs won't always be that convenient
    * e.g TfJob specs can be rather verbose
* Given the above, I think there is an opportunity to build higher level libraries and tooling ontop of these lower APIs that are easier to consume and perhaps hide K8s APIs.
* Overall I think this is in line with what's diagrammed in Fig 4.
* I think of Kubeflow as the entire stack as opposed to a specific API or component within the stack.
    * So I'd suggest replacing the box "Kubeflow" in Figures 1 & 2. with one or more components that's more descriptive.
* Based on the discussion some of the components that I think should be added to the diagrams are
    * UI
    * Logs server
         * Provides a consistent API for fetching job logs irrespective of the backend (e.g. that a cluster is using).
    * Extended API Server for managing TfJobs and/or other resources
         * I think there's an open question whether the functionality K8s provides by default for every CRD is sufficient, or whether we need to add additional methods
         * This already came up to some extent in tensorflow/k8s#125 because that PR introduces apihandler.go to support the UI ([discussion](https://github.com/tensorflow/k8s/pull/125#discussion_r153539800))

@DjangoPeng Based on the above do you still think we need to introduce a Type to TfJob?

**Comment by ddysher** [comment]
Thanks for the write up!  

In general, I can only second on @jlewi's comment about having different proposals on different components; and most importantly, how they fit together.  It's not clear to me the connections between different components just from reading the proposal.  My impression is that we have two layers in kubeflow - one is the lower Kubernetes level, where CRDs are maintained (not just creating CRD, but actually maintaining CRD, like updating status); the other layer is a more user-friendly layer; that's where the concept of `server` in the proposal comes in, plus UI, CLI, advanced state monitoring, etc.  Initially, the two layers communicate through CRD APIs; later, we'll see if more options are needed.

Comments around different components:

- CRD: I've briefly scanned the [linked issue](https://github.com/tensorflow/k8s/issues/209), wondering why we use a single CRD for all types instead of three CRDs and one for each type?
- controller: It looks like controller is only used to create/deploy TF jobs? If so, who's responsible for managing the lifecycle of the TF jobs?  If the answer is monitor, then I guess my question is why it is put above Kubernetes Extended?
- Server:
  - requests from KubeFlow monitor for status updating <- why do you want `monitor` to post TFJob status update to `server`?  shouldn't it update status directly to CRD.status?
  - Recording status of TensorFlow jobs in order to reschedule TensorFlow jobs <- we want to make server stateless, so no `Recording` here I guess
- Monitor: from the description, seems `monitor` can be part of `controller`?  `controller` plus `monitor` maintains TF job lifecycle, which is exactly what a Operator meant to do.

Ignore my question if we have more detailed proposal coming to address them.

Also, I do have some other suggestions:

- As kubeflow is in a state of flux right now, we want to make sure the scope of the proposal is well-understood.  A non-goal section is preferred here.  For example, based on the description of `workflow`, it seems the proposal only cares about individual workflow, instead of a full workflow engine (e.g. from code to serving). Although we are considering tools around that, e.g. argo, if it's out of scope of this proposal, then putting a non-goal here is much clearer :)
- Goals are too specific and sometimes include details about how you plan to implement them. For example, the goal below doesn't tell why we need a monitor at first place :)
> Implement a monitor, to get the job status from Kubernetes and report specific status to the KubeFlow server`.

**Comment by Jimexist** [comment]
I want to second @ddysher's call for:
> A non-goal section is preferred here. 

Specifically, these parts are not in my understanding of the scope of KubeFlow:
- a job queue, within which each item is a request (e.g. to be submitted to the aforementioned _server_) for creating a TfJob
- user-facing multi-user or multi-tenant management (and their state persistence, APIs, etc.)

👆 what do you think?

**Comment by erikerlandson** [comment]
@jlewi @gaocegege 
My thinking is that CRDs themselves ought to be able to manage lifecycle state tracking. So in that sense etcd would be the "db".

**Comment by erikerlandson** [comment]
@Jimexist agreed on multitenancy.  With respect to "job queue" I'm not sure a full-blown job mgmt queue is in scope, but some kind of solution will be needed to ensure that when a TF job is scheduled, all the necessary pods will be able to run. An admission controller, at least, seems like a requirement, although it could be decoupled from kubeflow itself.

**Comment by erikerlandson** [comment]
A couple of meta-level comments:
1) kubeflow ought to be moved to a community such as `https://github.com/kubernetes/kubeflow`
2) These proposals might be better served as g-docs

**Comment by erikerlandson** [comment]
I agree w/ @ddysher 's point on multiple CRDs vs a single CRD.  Among other things, that would allow for finer granularity of submission policies as a function of the underlying tools/algorithms. For example, jobs that require _all_ pods to run can be distinguished from jobs that may not.

**Comment by elmiko** [comment]
my impressions from the BoF and Kubeflow sessions at KubeCon were that initially this project was going to curate CRDs and Kubernetes resource files for various types of frameworks and processing platforms, but this proposal seems like we are diving into some sort of managed application for running TensorFlow.

it would be really helpful to understand what the overall goals of the Kubeflow project are going to be, i realize it's a little amorphous currently but are we looking to start with a solid foundation and then build applications on top of that, or jump in with some sort of managed application right from the start, or maybe something in-between?

imo, starting off by solidifying the CRDs and whatever controllers might be created, as well as defining clear processes and standards for adding new resources to this repo seem like they would really help out our efforts to build here.

i'm wondering if anyone else has had thoughts along these lines?

**Comment by aronchick** [comment]
@erikerlandson We've thought about moving the stuff to a kubernetes community, but the cost is very high. That's not to say we don't want to! We're looking for the best avenue to achieve this quickly, with low friction on the community. Is the concern that it's under Google or something else? It's all Apache2, so we (Google) couldn't do anything like lock down the project (people could just fork if we did).

In re: @ddysher's point about multiple vs. single CRDs, we've heard from both sides here. Can we create a proposal/issue to discuss that on?

In re: @erikerlandson's point about proposals, I completely agree. I'd like to propose this document is the starting point for an overall architecture documentation/flow for Kubeflow, and the items here are broken out into individual proposals in the GH repo. This issue is already hideously big :)

In re: @elmiko's point about a managed app for TF, that's definitely not the case. We very much want to work with many different open source projects including other serving tech, ML frameworks, things that are completely outside this scope (e.g. via connectors), and so on. But, you're exactly right, we need to get the core of the platform working great first. We started with Jupyter, TF CRD and TF serving because that's what we know, but we'd love to extend as fast as humanly possible (when others who know the other tools can plug in).

**Comment by foxish** [comment]
I don't think this is a great fit for a k8s sub-project. See https://github.com/kubernetes/community/blob/master/incubator.md#applications-on-top-of-kubernetes. +1 for proposals as gdocs.  Job management, gang scheduling, etc are concerns for the broader k8s project and not just limited to kubeflow, so, it makes sense to solve them outside this kubeflow in a more general way.

**Comment by erikerlandson** [comment]
Regarding the cost of moving the project, any resulting friction from moving it seems likely to only increase over time.

It could be ported into a new GH org like so:
1. create GH org `https://github.com/kubeflow`
1. fork repo `https://github.com/google/kubeflow` into org `https://github.com/kubeflow`
1. decomission repo `https://github.com/google/kubeflow`
1. move forward with repo (and org) `https://github.com/kubeflow/kubeflow`

We did something very similar with spark-on-kube about a year ago, and it has been a great platform for growing a dev and user community around.  Pragmatically it's very convenient to on-board new org owners or committers, in a way that is decoupled from any particular company. It provides a foundation for community governance.

Just as a POC, I just created an org `kubeflowio` (kubeflow is taken) and forked the repo:
https://github.com/kubeflowio/kubeflow

**Comment by aronchick** [comment]
@erikerlandson Can you specify your issue with hosting under Google _for now_?

I own the Kubeflow repo, the tech behind moving it is incredibly trivial, and would not cost us anything (almost ever). I was there when we moved Kubernetes from Google to Kubernetes, for example :) 

The "cost" I'm concerned with is the tax of being part of Kubernetes or CNCF or any of the above. They have much steeper requirements than we do in this repo, and I want to make that a thoughtful decision. As @foxish points out, a sub-project in K8s doesn't make sense right now, and since it doesn't APPEAR to be a concern for anyone, why not take the easy route? But I'm more than open to hearing if I'm missing something!

**Comment by foxish** [comment]
I'm with @aronchick on this. I think the case with Spark was different in that we moved from a private repo to a common location. This is already open and under Apache2 for contributions, (not to mention also having 1000+ stars and lots of watchers, and some press already) - moving it now would not give us any additional benefits as far as I can see.

**Comment by erikerlandson** [comment]
I think it's totally sensible to create an org that is not affiliated with kube, CNCF or any other org.  My main concern with "we can do it later" is that I expect this project to grow rapidly and the gravity-well around the current url will grow correspondingly fast.

I fundamentally agree that the technical differences are not large. I propose that there will be symbolic and psychological value. I would also expect that to translate into other more concrete benefits in terms of community engagement.

If the longer-term goal *is* to attach to some other project (e.g. CNCF) then coming from a position as an independent OSS community org can't hurt and might help.

**Comment by willb** [comment]
@aronchick @foxish I agree with Erik's sentiment that there's tremendous value in publishing open-source projects outside of company-specific organizations.  It shows that a project is intended to be a community, not merely a repository of permissively-licensed code.  It's hard to understand how having a project under a company's brand can chill community engagement until you've seen it play out.  (One example:  at Red Hat, we've even been surprised by the perception that some projects on fedorahosted.org are "just Red Hat code" -- and the majority of contributors to the Fedora community aren't employed by Red Hat!)  GitHub organizations are free -- why not use one that reflects the goal of a broad and engaged community?  (I get that having the URL in the wild already complicates things, but those watchers and stargazers will move with the code when you transfer a repo between organizations.)

**Comment by aronchick** [comment]
Again, my point isn't that we shouldn't do it - my point is that no one has brought this up as a reason not to contribute _now_.

For example, Draft is still under Azure, and has great momentum. RedHat hasn't contributed OpenShift to a foundation. And so on.

All I'm saying is let's wait a bit! There's no rush, and we haven't had any chilling effect that wasn't imminently addressable. Further, because we're under Google we get a _LOT_ of "free stuff" - attention, interest, contributions, etc. Are we sure that moving helps more than it hurts?

I am more than ready to have a conversation with ANYONE who is chilled by the current status and about their concerns. Kubernetes was under Google for more than 18 months, and it didn't hurt the growth at all.

**AGAIN, JUST TO REITERATE** I am totally and completely open to moving it, I just want to do it thoughtfully.

**Comment by willb** [comment]
I totally appreciate the discussion here, @aronchick!  Thanks for listening to these concerns.

Regarding chilling effects, the issue isn't the conversations we do have in open-source communities, it's about the conversations people never start.  Again, from experience, people who are chilled by contributing to repositories hosted under corporate organizations have to be pretty comfortable with someone involved with the project before they'll even raise the issue.  

Regarding potential downsides to a different organization, it is obvious that this repository has gotten a great deal of attention because of its association with Google.  But it's not clear that it would be any less obvious that the genesis of this project was at Google if it were hosted in a different GitHub organization.  (We should also consider the goal of open-source work:  do we wish merely to accumulate engagement or to attract a diverse group of contributors and build sustainable, healthy communities?)

Finally, regarding the OpenShift analogy, we're talking about GitHub organizations and repository URLs.  We aren't talking about foundations yet, and I agree that the choice of a foundation is a bigger decision.  (But that also points to one of the several reasons why OpenShift is a bad analogy, since it isn't hosted under a Red Hat GitHub organization.  Indeed, there _isn't one_ -- github.com/redhat isn't associated with Red Hat, Inc. at all.)

**Comment by aronchick** [comment]
> We should also consider the goal of open-source work: do we wish merely to accumulate engagement or to attract a diverse group of contributors and build sustainable, healthy communities?

This is a really insightful comment and the above, I think, _really_ gets at the core. We should decide this together! May I propose at some point (do we have enough quorum to do this now? maybe in the new year?), we get an hour with whomever would like to join a call and discuss our overall goals. Then we can make sure we choose the right structure to achieve these goals.

If that works for everyone, I'll propose a time on kubeflow-discuss@ - we have some interest from some first class community managers who may be able to help, and I'd like to get their thoughts on how to kick that off. 

Thoughts?

**Comment by jlewi** [comment]
@aronchick Can you open a separate issue(s) to discuss community organization issues? And lets use this issue to continue discussing the proposal from @DjangoPeng

**Comment by aronchick** [comment]
I would propose closing and forking this issue into its component parts. It's got way too much included!

**Comment by ddysher** [comment]
@aronchick How about moving this into a gdoc so folks interested in the topic can continue their discussion?  Closing the issue is fine but I would like to have a place where we can discuss some high level design/architectures of kubeflow to governing the discussion of specific components.

**Comment by DjangoPeng** [comment]
Agree with @ddysher , I have move the proposal into a gdoc [here](https://docs.google.com/document/d/1dmErPUmqqKMOe4L0ZHQglSdgDguCM4SzlsEdYXRMIDA/edit?usp=sharing). Welcome to continually discuss the topic in the gdoc.

@jlewi I think we don't need the type field if we define multiple CRDs.

**Comment by klaus82** [comment]
Hello all,
I think this project is a great idea.
What can I do to help you?

**Comment by aronchick** [comment]
Hi Claudio!

My best recommendation, for right now, would just to be to try out the
project in whatever way you'd like (e.g. on GKE, on AWS, on your laptop,
whatever). As you go through that process, try to write down areas where it
wasn't as smooth as you'd like, and file those as bugs. Or, if you'd like,
after you file them, start picking them up to improve them!

Thank you so much for volunteering!

**Comment by klaus82** [comment]
Thank you David!
I'll do for sure!

**Comment by jlewi** [comment]
Closing the thread per discussion above.

The proposal has moved to a [Google doc](https://docs.google.com/document/d/1dmErPUmqqKMOe4L0ZHQglSdgDguCM4SzlsEdYXRMIDA/edit)

There is a [thread](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubeflow-discuss/hMpk9uSqlIw/l_fUqiDkBgAJ) in kubeflow-disccuss to organize a meeting discuss community organization.

@ddysher Maybe send a PR adding a link to your doc to the README to make it more discoverable?

**Comment by ddysher** [comment]
@jlewi do you mean the [Google doc](https://docs.google.com/document/d/1dmErPUmqqKMOe4L0ZHQglSdgDguCM4SzlsEdYXRMIDA/edit#) version of the issue?

**Comment by jlewi** [comment]
Yes.

**Comment by jlewi** [comment]
@elgalu I think we are still working on trying to resolve the experiment tracking issue see

#136 Model Management Features
kubeflow/community#195
